File: python/euler/bin/p0052.py
# project euler: problem 52

from itertools import count


def check_num(n: int) -> bool:
    def get_key(num: int) -> str:
        return ''.join(sorted(str(num)))

    key_id = get_key(n)

    return all(key_id == get_key(n * mult) for mult in range(2, 7))


def compute() -> str:
    for ndigits in count(6):
        lower = 10 ** (ndigits - 1)
        upper = (10**ndigits) // 6
        for n in range(lower, upper + 1):
            if check_num(n):
                return str(n)

    raise RuntimeError('unreachable!')


def solve() -> str:
    return compute()

File: typescript/euler/bin/p0052.ts
// project euler: problem 52

import { range } from "../lib/util.ts";

const checkNum = (n: number): boolean => {
  const getKey = (n: number): string => String(n).split("").sort().join("");

  const keyId = getKey(n);
  for (const mult of range(2, 7)) {
    if (keyId !== getKey(n * mult)) {
      return false;
    }
  }

  return true;
};

export const compute = (): string => {
  let ndigits = 6;
  while (true) {
    const lower = 10 ** (ndigits - 1);
    const upper = Math.trunc((10 ** ndigits) / 6);
    for (const n of range(lower, upper + 1)) {
      if (checkNum(n)) {
        return String(n);
      }
    }
    ndigits += 1;
  }
};

export const solve = (): string => compute();
